Introduce SeleniumBrowser

signalprime opened the pull request:
### Key Contributions:
- **SeleniumBrowser Function**: 
Adds a headless, fully-functional desktop web driver to enable dynamic interactions with web pages, crucial for accessing content reliant on JavaScript or other client-side scripts.

- **SeleniumBrowserWrapper Class**: 
Offers a seamless alternative to the `SimpleTextBrowser`, enhancing agent capabilities in interacting with web pages that require advanced browsing functionalities.

- **WebSurferAgent Enhancements**: 
This update enables the ability to select between `SimpleTextBrowser` or `SeleniumBrowserWrapper` based on the provided `browser_config` and opens the door to vision-based function calling and interactions, significantly broadening the scope of tasks and potential use cases for Autogen agents in web interaction scenarios going forward.  

- **Unit Testing**: 
Extends unit tests to cover the enhanced `WebSurferAgent`, ensuring both functionality and reliability in the agents' operations.

- **Notebooks**:
Added `agentchat_surfer_edge.ipynb` to demonstrate cross compatibility and new graphical functionality
Both notebooks specify `gpt-3.5-turbo` and rely on round-trip generations without follow-ons.

### Benefits
- Broadens the scope of web-based content that can be collected
- Facilitates future development of sophisticated agents capable of complex web-based vision tasks
- Enhances the WebSurferAgent with configurable browser behavior

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR. *need assistance to tell GitHub workflows that this PR requires websurfer extras*
- [x] I've made sure all auto checks for `WebArchiverAgent` have passed. 
- [x] I've made sure all auto checks for `WebSurferAgent` have passed.

signalprime commented:
@microsoft-github-policy-service agree


On Mon, Feb 19, 2024 at 10:24 PM Codecov Comments Bot <
***@***.***> wrote:

> Codecov
> <https://app.codecov.io/gh/microsoft/autogen/pull/1733?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft>
> Report
>
> Attention: 396 lines in your changes are missing coverage. Please review.
>
> Comparison is base (2750391)
> <https://app.codecov.io/gh/microsoft/autogen/commit/2750391f847b7168d842dfcb815ac37bd94c9a0e?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft>
> 39.33% compared to head (3954412)
> <https://app.codecov.io/gh/microsoft/autogen/pull/1733?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft>
> 21.53%.
>
> Files
> <https://app.codecov.io/gh/microsoft/autogen/pull/1733?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft> Patch
> % Lines
> autogen/agentchat/contrib/content_agent.py
> <https://app.codecov.io/gh/microsoft/autogen/pull/1733?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft#diff-YXV0b2dlbi9hZ2VudGNoYXQvY29udHJpYi9jb250ZW50X2FnZW50LnB5>
> 0.00% 208 Missing and 3 partials ⚠️
> <https://app.codecov.io/gh/microsoft/autogen/pull/1733?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft>
> autogen/browser_utils.py
> <https://app.codecov.io/gh/microsoft/autogen/pull/1733?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft#diff-YXV0b2dlbi9icm93c2VyX3V0aWxzLnB5>
> 4.76% 179 Missing and 1 partial ⚠️
> <https://app.codecov.io/gh/microsoft/autogen/pull/1733?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft>
> autogen/agentchat/contrib/web_surfer.py
> <https://app.codecov.io/gh/microsoft/autogen/pull/1733?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft#diff-YXV0b2dlbi9hZ2VudGNoYXQvY29udHJpYi93ZWJfc3VyZmVyLnB5>
> 44.44% 5 Missing ⚠️
> <https://app.codecov.io/gh/microsoft/autogen/pull/1733?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft> Additional
> details and impacted files
>
> @@             Coverage Diff             @@##             main    #1733       +/-   ##
> ===========================================- Coverage   39.33%   21.53%   -17.81%
> ===========================================
>   Files          57       58        +1
>   Lines        6096     6502      +406
>   Branches     1365     1564      +199
> ===========================================- Hits         2398     1400      -998     - Misses       3502     4932     +1430     + Partials      196      170       -26
>
> Flag
> <https://app.codecov.io/gh/microsoft/autogen/pull/1733/flags?src=pr&el=flags&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft> Coverage
> Δ
> unittests
> <https://app.codecov.io/gh/microsoft/autogen/pull/1733/flags?src=pr&el=flag&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft> 21.53%
> <3.17%> (-17.81%) ⬇️
>
> Flags with carried forward coverage won't be shown. Click here
> <https://docs.codecov.io/docs/carryforward-flags?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft#carryforward-flags-in-the-pull-request-comment>
> to find out more.
>
> ☔ View full report in Codecov by Sentry
> <https://app.codecov.io/gh/microsoft/autogen/pull/1733?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft>
> .
> 📢 Have feedback on the report? Share it here
> <https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=microsoft>
> .
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/autogen/pull/1733#issuecomment-1953467303>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADWFCMFGKUPKH3OVLMHL573YUQQRDAVCNFSM6AAAAABDQOESUOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNJTGQ3DOMZQGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

signalprime commented:
This PR expands the `websurfer` extras requirements to include `selenium`.  I've also updated the workflow for `python-package.yml` to include the webdrivers for Edge, Firefox, and Chrome.  

All `pre-commit` tests in the dev docker instance pass, with one exception due to `no-commit-to-branch`.  I realize after the fact that committing to a separate branch would have been preferred and was able to able to commit using `SKIP=no-commit-to-branch git commit ...`.

Finally, regarding the `build` workflow, there are two pending issues:
1. It is still unclear how to inform the system that it should install the `websurfer` extra requirements for this PR
2. We're getting numerous `PytestUnknownMarkWarning: Unknown pytest.mark.asyncio` errors, however that markdown is used throughout and I wonder if it's an issue regarding the installed version of pytest.  As a side note in reporting, in the dev docker I had to upgrade pydantic to 2.5.3 when it couldn't find `field_validator` in a file unrelated to this PR

thinkall commented on .github/workflows/python-package.yml:
```diff
@@ -54,3 +54,15 @@ jobs:
           TWINE_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
         shell: pwsh
         run: twine upload dist/*
+
+      - name: Setup edge
+        uses: browser-actions/setup-edge@latest
+        if: ${{ matrix.browsers == 'edge' }}
+
+      - name: Setup firefox
+        uses: browser-actions/setup-firefox@latest
+        if: ${{ matrix.browsers == 'firefox' }}
+
+      - name: Setup chrome
+        uses: browser-actions/setup-chrome@latest
+        if: ${{ matrix.browsers == 'chrome' }}
```
The test settings for WebSurfer is here https://github.com/microsoft/autogen/blob/main/.github/workflows/contrib-tests.yml#L183-L218

signalprime replied:
Thank you so much for your guidance with this @thinkall, very much appreciated! :)

signalprime replied:
I moved the browser actions as we discussed.  The build tests are failing with the error [ModuleNotFoundError: No module named 'markdownify'](https://github.com/signalprime/autogen/blob/25e15e05cc95af0bad97c6cbc0324f4b3c259258/autogen/browser_utils.py#L6)

I confirmed `markdownify` is installed with the [`websurfer` requirements](https://github.com/signalprime/autogen/blob/25e15e05cc95af0bad97c6cbc0324f4b3c259258/setup.py#L56) [here](https://github.com/signalprime/autogen/blob/25e15e05cc95af0bad97c6cbc0324f4b3c259258/.github/workflows/contrib-tests.yml#L202) and [here](https://github.com/signalprime/autogen/blob/25e15e05cc95af0bad97c6cbc0324f4b3c259258/.github/workflows/contrib-openai.yml#L245), just before the browser actions are specified.  

I will research more to see how I can solve it.

signalprime replied:
@afourney hi friend, if I'm not mistaken, you introduced the WebSurfer agent.  Could you offer any additional advice about how I could get this PR to trigger the websurfer extras installation during build tests and avoid import errors for libraries like `markdownify`?

signalprime commented:
- Updated workflows to include the browsers, I'll review again to ensure those are all configured
-- contrib-openai.yml 
-- contrib-tests.yml 
- Updated `setup.py` so that `websurfer` includes `selenium`, `requests`, and `arxiv` packages
- Added additional typing and docstrings to `browser_utils.py`
- Updates to `SeleniumWrapper` to ensure 100% coverage and functionality seen in `SimpleTextBrowser`
- Added `notebook/agentchat_surfer_edge.ipynb` to demonstrate cross compatibility as well as new graphical functions 
- Confirmed functionality with Bing API
- Added `notebook/agentchat_content_agent.ipynb` to demonstrate core functionality
- pre-commit updated unrelated notebooks `notebook/agentchat_lmm_gpt-4v.ipynb` and `notebook/agentchat_custom_model.ipynb` to my surprise, removal or addition of a empty line(s)

signalprime commented on autogen/browser_utils.py:
```diff
@@ -280,3 +313,741 @@ def _fetch_page(self, url: str) -> None:
+        webdriver: An instance of the Selenium WebDriver based on the specified browser.  User can open a new page by `webdriver.get('https://www.microsoft.com')`.
+
+    Raises:
+        ImportError: If selenium package is not installed, it raises an ImportError with a message suggesting to install it using pip.
```
This no longer raises an ImportError because PEP 8 guidelines prevent imports that are not at the start of the file.  I should remove this "Raises:" comment in the docstring because we have handled this during the conditional statement of whether to use the `SeleniumBrowserWrapper` or not.

signalprime commented on autogen/browser_utils.py:
```diff
@@ -280,3 +313,741 @@ def _fetch_page(self, url: str) -> None:
+    if not download_dir:
+        download_dir = tempfile.gettempdir()
+
+    browser_res = kwargs.get("resolution", (1920, 5200))
```
Recent updates for screenshots either resize the window to fit the rendered content or, in the case of Firefox, use the full page screenshot.  It may be more memory efficient for all users to have a standard 1080 height.

signalprime commented on autogen/browser_utils.py:
```diff
@@ -280,3 +313,741 @@ def _fetch_page(self, url: str) -> None:
+        # Instantiate the Firefox WebDriver with the configured options
+        driver = webdriver.Firefox(
+            options=get_headless_options(download_dir, options)
+        )  # , service_log_path=f'{tempfile.tempdir}/geckodriver.log')
```
I can delete this extra comment, and a few others above it.  They are remnants from ensuring coverage during the graphical websurfer notebook.

signalprime commented on autogen/browser_utils.py:
```diff
@@ -280,3 +313,741 @@ def _fetch_page(self, url: str) -> None:
+        # self._page_content # or self.driver.page_source
+        # Image.open(io.BytesIO(self.driver.get_screenshot_as_png()))
+        # if self._page_content and len(self._page_content) > 0
+        return self._page_content
```
Same as above, I can remove these comments.  At this point, we have a function to show what is on the screen using a screenshot to GUI or screenshot to disk.  The page content is always text-based which ensures compatibility with established WebSurfer functionality

signalprime commented on autogen/browser_utils.py:
```diff
@@ -280,3 +313,741 @@ def _fetch_page(self, url: str) -> None:
+
+    def download(self, uri_or_path: str) -> None:  # TODO: update this based on the new method
+        """Download from a given URI"""
+        self.driver.get(uri_or_path)
```
Update this to use `download_using_requests` from browser_utils.py

signalprime commented on notebook/agentchat_custom_model.ipynb:
```diff
@@ -383,6 +383,7 @@
     "# load model here\n",
     "\n",
+    "\n",
     "config = config_list_custom[0]\n",
```
My apologies, this and one other change were made by pre-commit checks and added inadvertently.  I'll try again to revert them and remove the two unrelated changes like this.

signalprime commented on test/agentchat/contrib/test_web_surfer_selenium.py:
```diff
@@ -0,0 +1,179 @@
+        assert f"Address: {BLOG_POST_URL}".strip() in response
+        assert f"Title: {BLOG_POST_TITLE}".strip() in response
+
+        if not skip_bing:
```
I can remove this since it is handled by pytest.mark.skipif

skzhang1 commented on autogen/agentchat/contrib/content_agent.py:
```diff
@@ -0,0 +1,483 @@
+    pass
+
+
+class ContentAgent(ConversableAgent):
```
ContentAgent is not informative. how about "WebArchiverAgent"?

signalprime replied:
I like that too.  I'd gone through a few ideas like WebContentAgent, and I like your suggestion.   Also thank you so much for reviewing my PR :)

skzhang1 commented on autogen/agentchat/contrib/content_agent.py:
```diff
@@ -0,0 +1,483 @@
+            f"Success: archived the following links in your chosen location {self.local_dir}/ <-- {', '.join(all_links)}",
+        )
+
+    def fetch_content(self, link):
```
fetch_content ---> _fetch_content

skzhang1 commented on autogen/agentchat/contrib/content_agent.py:
```diff
@@ -0,0 +1,483 @@
+        else:
+            return "html", self.fetch_html_content(link)
+
+    def fetch_html_content(self, link):
```
fetch_html_content -----> _fetch_html_content. Change the remaining private methods

skzhang1 commented:
Thank you for the PR! In my opinion, it would be good to include some code examples in the notebook to demonstrate the advantages of the ContentAgent. From what I understand, the ContentAgent is a method for aggregating web information onto disk. However, I’m not sure whether the original assistant and userproxy agents could achieve the same goal through a meticulously designed prompt.

signalprime commented:
> I’m not sure whether the original assistant and userproxy agents could achieve the same goal through a meticulously designed prompt.
Yes, I think you are correct that it could conceptually be handled strictly using existing agents and prompts.  I did it this way for two reasons.  The first being that this agent works as part of a larger pipeline that will require the archived content, and the second reason is from an energy conservation perspective.  Hopefully with recent developments it won't be a concern, but for now we ask if it's responsible to take the tractor (LLM inference) to pickup more cat food or just ride our bike (a scripted series of functions to archive the components).   That said, an LLM when done right could potentially handle the corner cases.  

The WebSurfer upgrades are a byproduct from the requirements of the WebArchival (thanks @skzhang1) agent, which is a dependency of my next PR.  I _could_ technically remove the Content or WebArchival Agent components from this PR.  I'm open to all ideas, especially what is best for Autogen as a whole.

signalprime commented:
Made some last updates based on the generous feedback from @skzhang1.  The only two other things I can think to include would be an update to the docs for the graphical WebSurferAgent and additional tests to `test/test_browser_utils.py`.  

@sonichi thank you for running the OAI tests.  I noticed `test_web_surfer.py` failed with a [NameError](https://github.com/microsoft/autogen/actions/runs/8039662927/job/21956892152#step:5:80) about `WebSurferAgent` being undefined.  That test file is unchanged in this PR, but the agent has small changes.  Looking into it we do see the agent class imported [here](https://github.com/signalprime/autogen/blob/451405b3156169f78a9d1e8831187d516da574d6/test/agentchat/contrib/test_web_surfer.py#L20) with a catch to disable all tests in the case of failure.  Your thoughts?

sonichi commented on autogen/browser_utils.py:
```diff
@@ -26,6 +34,31 @@
+
+    IS_SELENIUM_CAPABLE = True
+except ImportError as e:
+    print(f"The module/package '{e.name}' is not available.")
+    print("Try running 'pip install selenium'.  You may need to run 'sudo easy_install selenium' on Linux or MacOS")
+    print(
+        "Official selenium installation documentation: https://www.selenium.dev/documentation/webdriver/getting_started/install_library/"
+    )
+    raise e
```
We probably shouldn't print these messges here. Nor raise the error. Only print and raise when selenium is required.

signalprime replied:
Thanks for your feedback, will remove it

sonichi commented on autogen/agentchat/contrib/web_surfer.py:
```diff
@@ -85,6 +106,18 @@ def __init__(
+
+    @property
+    def render_text(self):
+        self.browser._set_page_content(self.browser.page_content)
```
can we avoid using private methods here?

signalprime replied:
I'll find a way to do it, updating coming Friday

sonichi commented on .github/workflows/contrib-openai.yml:
To run this test, please change the `pull_request_target` into `pull_request`.

sonichi reviewed (commented):
I'd like to get @afourney 's opinion on the structure of the changes, e.g., whether it's ok to directly modify the current web surfer agent or should a separate agent be created. And whether a separate optional dependency is desired. My other comments may not apply if a structural change is needed.

signalprime commented:
> I'd like to get @afourney 's opinion on the structure of the changes, e.g., whether it's ok to directly modify the current web surfer agent or should a separate agent be created. And whether a separate optional dependency is desired. My other comments may not apply if a structural change is needed.

Hi @afourney, I saw in an issue that you were thinking to pass a browser object to the agent.  Would you prefer that to the method I have configured?  I could expose an option to provide a browser object while maintaining the current behavior where the browser type is specified in the config to keep things simple for the avg end-user.  Your thoughts?

signalprime commented:
I don't know why @afourney chose not to reply here, but 2 weeks later began introducing new PRs without accepting mine first.  I'd suggested to accept my contribution and then do whatever he wants afterwards.  That appeared to clearly be the best way to proceed.

MohammedNagdy commented:
@signalprime i would like you to take a look at my [PR](https://github.com/microsoft/autogen/pull/3049) maybe we can adopt one design as I’m working on a similar part.

rysweet commented:
This PR is against AutoGen 0.2. AutoGen 0.2 has been moved to the 0.2 branch. Please rebase your PR on the 0.2 branch or update it to work with the new AutoGen 0.4 that is now in main.

rysweet commented:
@signalprime unfortunately after rebase there are still some open conflicts. If you are still interested in bringing this one forward please see if you can get those resolved and green the latest CI.

rysweet commented:
closing as stale, please reopen if you would like to update
